### ISSUE TITLE
Memory leaks when using a Flutter project

### ISSUE THREAD

**Issue by Skylled** [Dart-Code/Dart-Code#964]
Normal coding use of Dart-Code seems to slowly increase memory usage.

The leak is not related to building or debugging an app.

As mentioned in Gitter, the Extract Widget test on loop is a good repro.

**Comment by DanTup** [comment]
I'm struggling to repro (I can balloon the memory, but generally only in ways that cleans up). Next time you see this, can you try closing all open files and then waiting for 3 minutes, and see if it's come down?

VS Code only cleans up documents every 3 mins, so if the memory goes back down, it's probably something attached to the editor.

Looking at the code, I think we could also be holding references because we don't clean up the handlers after getting responses from the analysis server - this seems like it should leak, but cleaning it up didn't seem to make any difference in my testing.

A heap snapshot might be useful, but honestly I'm not sure I know how to read them in Chrome - I've been digging through one just now and I can't really figure out how to track the instances back to anything useful! :/

**Comment by DanTup** [comment]
@Skylled How easily can you repro this? Can you do it with other extensions disabled? I've been using it for a while, including opening lots of files, opening completions lots, etc.; but my memory usage is still only a few hundred MB :-(

I can definitely thing of some places we might not be cleaning up very well (and I'm gonna fix), but I don't know if any of these are what you're hitting.

**Comment by DanTup** [comment]
Here's what I see after loads of opening/fake editing:

![screen shot 2018-06-04 at 12 28 19 pm](https://user-images.githubusercontent.com/1078012/40915134-cdd7e210-67f2-11e8-9c0d-0826b4ff7e20.png)

**Comment by DanTup** [comment]
Ok, I can get the memory up a little by opening all of the flutter `example` projects as a multi-root workspace and then opening the Workspace Symbols list and typing/deleting characters repeatedly. Memory usage can climb up to 1GB; however it does come back down on its own shortly after, so I think that might be just be because of the huge number of results we're creating and throwing away.

I'm gonna see if I can cause leaks in the things I suspect are leaky (thread cleanup on debugging and keeping hold of handlers for requests to the analysis server) by running code in a loop and address them at least.

**Comment by Skylled** [comment]
If I had to guess what the issue was related to, I'd guess code completion, as that's the first way that I notice the problem: slow response time from code completion. But that could simply be a side effect of the memory.

When next I have time to work, I'll try Code's extension debugging. Maybe there's a number of exceptions or warnings being hidden from the user.

If you have any specific extra steps you would like me to try, please let me know.

**Comment by DanTup** [comment]
I've made a change that cleans up the handlers for requests to the server, but I think at worse that would hold on to things in the request, not the response; so I'm not sure if that helps. Still, I'll run it a bunch in a loop and see what turns up.

If you can reliably repro (even if it means working for a few hours) I can provide a v2.14 alpha build soon that has that change if you want to see if it makes a different.

**Comment by Skylled** [comment]
No need, I'll run from master branch while we work this out.

**Comment by DanTup** [comment]
Master is still v2.13; `dev` is v2.14 (will merge in a few days after v2.13.1 is out). These changes are currently on the `reduce-memory-usage` branch. Running completion in a loop is definitely blowing memory, just trying to figure out if it's leaking or just cleaning up slowly. I'll post back shortly...

**Comment by DanTup** [comment]
Opened an issue with VS Code to try and get some info on how best to profile this. I think the profiles I was looking at before were for the editor side, which is why nothing looked familiar. I'm unable to attach to the URL from the console in Chrome though :(

https://github.com/Microsoft/vscode/issues/51085

**Comment by DanTup** [comment]
Ok, I don't think any of my fixes so far will help. Completion isn't a request/response, the response just contains an ID which we use to then listen for a notification. I've checked, and we were already unsubscribing that handler when the results came in.

I really can't see anything in the completion provider that should be able to leak (though the JSON.parse does seem pretty slow because of the large number of results).

**Comment by Skylled** [comment]
Is there anything you can think of that would cause this to only happen in Flutter?

My vanilla Dart server code uses Flutter's SDK repo (no sense in duplicating the SDK on my drive), and I don't have memory issues there.

**Comment by DanTup** [comment]
@Skylled Nothing I can think of, and certainly not in the completion provider. The completion provider may return more stuff in a Flutter project though, so it could just be that it's happening much more slowly in Dart?

No response to my VS Code issue asking about getting heap snapshots :(

Maybe I should add a setting or something to log [memory usage](https://nodejs.org/api/process.html#process_process_memoryusage) periodically and we can spot any patterns.

You're sure it's happening with no debug sessions? Are you opening lots of files (or big files), making large edits, or anything else quite frequently?

**Comment by Skylled** [comment]
Correct. Consistently happens without debug sessions. I don't make big changes usually. I switched from having just one large file to maybe a dozen smaller ones.

I'm going to fiddle with my settings some to see if anything shakes loose, starting with format on type.

**Comment by DanTup** [comment]
Format-on-type only actually triggers on certain key presses (like `}`) so might not be triggering as much as you think. However, now you mention it - we do get large strings back from the server for formats (it doesn't support formatting small ranges) but I can't see anything obvious that'd leak there (except for references to the document, but I don't imagine that'd do much unless you had huge numbers of files).

`master` now has the leak fixes I did make; feel free to try with that - or I'll probably make a v2.14 alpha soon if you want to run with that to see whether it was fixed with the changes above?

**Comment by DanTup** [comment]
Ok, I can get the dev tools connected properly now. If you launch code like this:

```
code --inspect-plugins=9333 .
```

That'll open Code for the current folder with the debugger listening. If you open the dev console (Help -> Toggle Dev Tools) there's a URL that you can paste into Chrome, which will give you a connection. You can then click Memory and Heap Snapshot to get a snapshot of the heap.

Unfortunately, mine has been sat like this for some time:

![screen shot 2018-06-06 at 10 14 53 am](https://user-images.githubusercontent.com/1078012/41029070-ad271438-6972-11e8-93b1-ec7f1af3dde3.png)

I saw people online claiming this could take *hours* for snapshots of just a few hundred MB if there are lots of cycles. If you can get it to work, it might give us an idea of what's consuming the memory (and why it isn't being freed).

**Comment by DanTup** [comment]
Well, if I start without opening completion, it's really fast. If I open completion once first, then it's a little slow. Open it a few times, and it's SUPER slow. I'll leave it to finish and see if I can find anything.

**Comment by DanTup** [comment]
Ok, might have found something... 

![screen shot 2018-06-06 at 10 32 55 am](https://user-images.githubusercontent.com/1078012/41029945-060b8c6c-6975-11e8-8291-be485f4bc6d6.png)

This is a comparison of what was allocated between Snapshot 3 (where I'd only opened completion a few times) and Snapshot 4 (where I'd opened it a bunch more). There are 5k `ResourceConfig`s allocated and none freed. A lot of the other items in there seem to be held by these `ResourceConfig`s. Guess what we create *for every item in every completion request*?

```ts
const insertArgumentPlaceholders = config.for(document.uri).insertArgumentPlaceholders;
```

And the reason they're not being freed?

```ts
constructor(uri: Uri) {
		this.uri = uri;
		workspace.onDidChangeConfiguration((e) => this.loadConfig());
		this.loadConfig();
	}
```

We're setting up an onDidConfigChange that notifies us when the config changes, for this temporary ResourceConfig we're going to throw away. So the config changes listener is holding all these references to all of these config objects we're not using!

So, two things should help here:

1. Don't subscribe to config changes inside these classes - I don't believe they're ever held such that we need notification changes.
2. Stop creating them for the same document inside the loop!

**Comment by DanTup** [comment]
This looks much better. I'm pushing a fix out in v2.13.2 for this one since it's really serious and the fix is trivial. The other fixes mentioned above (and anything else I find with some further testing) will go into vNext.

Thanks!

**Comment by DanTup** [comment]
Fix is live in v2.13.2. Please update and let me know it your issues disappear. I'm also gonna do some additional profiling as part of #983 to ensure we don't have anything else similar.

**Comment by Skylled** [comment]
Finally got some time to check this, and yes, my memory usage has greatly decreased since this patch, which has improved the overall speed of Dart-Code. Thanks for this one!

**Comment by DanTup** [comment]
@Skylled Thanks for confirming! There are some other (more minor) leak fixes in v2.14, but if you see anything like this again; do shout me! :-)
